dig.go

```diff
@@ -253,13 +253,27 @@ type node struct {
 	provides reflect.Type
 	ctor     interface{}
 	ctype    reflect.Type
-	deps     []reflect.Type
+	deps     []dep
+}
+
+type dep struct {
+	Type     reflect.Type
+	Optional bool
 }
 
 func newNode(provides reflect.Type, ctor interface{}, ctype reflect.Type) (node, error) {
-	deps := make([]reflect.Type, 0, ctype.NumIn())
+	deps := make([]dep, 0, ctype.NumIn())
 	for i := 0; i < ctype.NumIn(); i++ {
-		deps = append(deps, getConstructorDependencies(ctype.In(i))...)
+		t := ctype.In(i)
+		if isInObject(t) {
```

Comment:
You can fold all this into the `getInDependencies` method, similar to `traverseOutTypes`.

Maybe even call it something similar as well, up to you